### PR TITLE
Fix segv while running camera_pipe

### PR DESCRIPTION
**Pull request by ronlieb**
seeing segv on a few tests, i think you might have to be running hexagon() to see the failures

Camera_pipe
Received archstring from environment=--l2cache_perfect 1
ARCHSTRING set l2cache_perfect = 1
make: *** [bin/out.png] Segmentation fault

a few correctness tests as well
bash: line 1: 47448 Segmentation fault      /local/mnt/workspace/bots/hexbotmaster-aus-perses-10-2/halide-82/src/halide/bin/correctness_argmax
make: *** [correctness_argmax] Error 139

### CODE REVIEW

File: src/runtime/device_interface.cpp [hunk 1]
```diff
@@ -324,7 +324,7 @@ WEAK int halide_device_and_host_free(void *user_context, struct buffer_t *buf) {
             } else {
                 return 0;
             }
-        } else {
+        } else if (buf && buf->host) {
```
Comment (dsharletg) [inline src/runtime/device_interface.cpp]: This function has a lot of issues, mostly preexisting:

- The debug statement dereferences buf even though it was checked for null-ness before (and after). 
- inconsistent conditions (buf and buf != NULL)
- Your new check checks buf is not null, but this is definitely true inside the outer if.

**Comment by ronlieb** [comment]
it started happening last mnight after this commit rolled into our builders

Change #479958

Category	downstream
Changed by	Andrew Adams <andrew.b.adams@gmail.com>
Changed at	Fri 24 Feb 2017 20:47:41
Repository	https://github.com/halide/Halide.git
Project	Halide
Branch	master
Revision	27bcb9c48d741586b89c7e76566ef829ab4653b4
Comments

Don't leak host memory when calling device_free on a combined allocation
It's possible for us to generate the following sequence of calls:

device_and_host_malloc
device_malloc
device_free
device_and_host_free_as_destructor

The device_malloc does nothing, then the device free frees the device
allocation only (on CUDA at least, I suppose it's up to the backend
whether or not this is a no-op). Normally there'd be a host free later
(a Free IR node), but suppose an error triggers before then - then
device_and_host_free_as_destructor is called via a register_destructor.
But this call does nothing if there's no device interface (because the
device interface implements device_and_host_free)! And there isn't,
because device_free was successful!

This was causing a nasty leak in the cleanup_on_error aottest. The test
was correctly failing, but the buildbots aren't running the generator
tests properly for GPU targets.
Changed files

src/runtime/device_interface.cpp
test/generator/cleanup_on_error_aottest.cpp

File: src/runtime/device_interface.cpp (hunk 1 above)
Comment (ronlieb) [inline src/runtime/device_interface.cpp]: i can change the else if to be 
   else if (buf->host) 
i am not familiar with this code to know if buf should have been non null and thus there is some issue elsewhere.

**Comment by ronlieb** [comment]
Dilllon, did you want me to submit a revised patch on this?

**Comment by abadams** [comment]
Sorry for the original bug. Was patching a memory leak. Should have gone through a PR.

**Comment by ronlieb** [comment]
No worries, thanks for pushing thru the fix, my bots will be happy now :)

 

 

---

Qualcomm Innovation Center, Inc. is a member of Code Aurora Forum, hosted by The Linux Foundation

**Pull request merged**